iModelCore/ECDb/ECDb/ClassViews.cpp

```diff
@@ -0,0 +1,194 @@
+    auto viewCA = viewClass.GetCustomAttributeLocal(ViewSchemaName, ViewClassName);
+    if (!viewCA.IsValid())
+        return false;
+
+    ECN::ECValue v;
+    if (ECObjectsStatus::Success != viewCA->GetValue(v, ViewClassQueryProp))
+        return false;
+
+    query = v.GetUtf8CP();
+    query.Trim();
+    if (query.empty())
+        return false;
+
+    return true;
+}
+
+//======================================================================================
+// @bsimethod
+//======================================================================================
+bvector<ECN::ECClassCP> ClassViews::FindViewClasses(ECDbCR conn) {
+    bvector<ECN::ECClassCP> viewClasses;
+    Statement stmt;
+    const auto rc = stmt.Prepare(conn, R"(
+        SELECT
+            [ca].[ContainerId]
+        FROM   [ec_class] [cs]
+            JOIN [ec_schema] [sc] ON [cs].[schemaId] = [sc].[Id]
+            JOIN [ec_CustomAttribute] [ca] ON [ca].[ClassId] = [cs].[Id]
+        WHERE  [ca].[ContainerType] = ?1
+                AND [sc].[Name] = ?2
+                AND ([cs].[Name] = ?3);
+    )");
+
+    if (rc != BE_SQLITE_OK) {
+        BeAssert(false && "unable to prepare statement");
+        return viewClasses;
+    }
+
+    stmt.BindInt(1, (int)SchemaPersistenceHelper::GeneralizedCustomAttributeContainerType::Class);
+    stmt.BindText(2, ViewSchemaName, Statement::MakeCopy::No);
+    stmt.BindText(3, ViewClassName, Statement::MakeCopy::No);
+    while(BE_SQLITE_ROW == stmt.Step()) {
+        const auto classId = stmt.GetValueId<ECN::ECClassId>(0);
+        const auto classP = conn.Schemas().GetClass(classId);
+        if (classP != nullptr) {
+            viewClasses.push_back(classP);
+        }
+    }
+    return viewClasses;
+}
+
+//======================================================================================
+// @bsimethod
+//======================================================================================
+bool ClassViews::CheckViews(ECDbCR conn) {
+    int invalidViews = 0;
+    auto dbViews = FindViewClasses(conn);
+    for (auto viewClass : FindViewClasses(conn)) {
+        if (!IsValid(*viewClass, conn)) {
+            ++invalidViews;
+        }
+    }
+    if (invalidViews > 0) {
+        conn.GetImpl().Issues().ReportV(
+                IssueSeverity::Error,
+                IssueCategory::BusinessProperties,
+                IssueType::ECDbIssue, "Total of %d view classes were checked and %d were found to be invalid.", (int)dbViews.size(), invalidViews);
+        return false;
+    }
+    return invalidViews == 0;
+}
+//======================================================================================
+// @bsimethod
+//======================================================================================
+bool ClassViews::IsValid(ECN::ECClassCR viewClass, ECDbCR conn){
+    auto viewClassName = viewClass.GetFullName();
+    if (!viewClass.IsEntityClass() && !viewClass.IsRelationshipClass()) {
```

Comment:
you allow the custom attribute to be applied to a relationship class but do not have a test when applied to a relationship class

Reply:
Yes, tests for relationships and a bunch of other stuff are still pending. Since now we allow struct/struct array etc we also need to test for those. We also do not support CTE in the query, but I have no check or test for it. We either must fix that or fail at it. We also missed the test of nonprimitive data return by one view that is referenced by another view.

I have not gotten to them yet. But here is definitely a lot of cases that are not covered yet.

Reply:
added linktable and nav property test.